### PR TITLE
feat(doc): set bold in selection

### PR DESCRIPTION
**Pull request by Jocs**
- [x] Set bold command
- [x] Undo/Redo set bold
- [x] Support multiple selections
- [x] Perceive the style in the selection, cancel it if it's already bold, and bold it if it doesn't

### CODE REVIEW

File: packages/base-docs/src/Controller/doc-render.controller.ts [hunk 1]
```diff
@@ -177,6 +177,8 @@ export class DocRenderController extends Disposable {
                         return;
                     }
 
+                    skeleton.calculate();
```
Comment (wzhudev) [inline packages/base-docs/src/Controller/doc-render.controller.ts:180]: 除了需要重新排版，由于设置或者取消粗体之后选中文字的宽度会发生变化，需要同时触发选区的重新渲染

File: packages/base-docs/src/Basics/commonParameter.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class CommonParameter {
```
Comment (wzhudev) [inline packages/base-docs/src/Basics/commonParameter.ts:1]: Maybe it deserves a better name. 🤣

File: packages/base-docs/src/Controller/inline-format.controller.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+} from '@univerjs/core';
+import { Inject } from '@wendellhu/redi';
+
+import CommonParameter from '../Basics/commonParameter';
+import { SetInlineFormatBoldCommand, SetInlineFormatCommand } from '../commands/commands/inline-format.command';
+import { IRichTextEditingMutationParams, RichTextEditingMutation } from '../commands/mutations/core-editing.mutation';
+import { TextSelectionManagerService } from '../services/text-selection-manager.service';
+
+@OnLifecycle(LifecycleStages.Rendered, InlineFormatController)
+export class InlineFormatController extends Disposable {
```
Comment (wzhudev) [inline packages/base-docs/src/Controller/inline-format.controller.ts:22]: Give it a simple description of its responsibility.

File: packages/base-docs/src/commands/mutations/core-editing.mutation.ts [hunk 4]
```diff
@@ -59,10 +67,18 @@ export const RichTextEditingMutation: IMutation<IRichTextEditingMutationParams,
 
         const undoMutations: Array<IRetainMutationParams | IInsertMutationParams | IDeleteMutationParams> = [];
         const commonParameter = new CommonParameter();
+
         commonParameter.reset();
+
         mutations.forEach((mutation) => {
+            // FIXME: @jocs Since UpdateAttributeApply modifies the mutation(used in undo/redo),
```
Comment (wzhudev) [inline packages/base-docs/src/commands/mutations/core-editing.mutation.ts:74]: Mutations should be frozen.

File: packages/core/src/Docs/Domain/DocumentBodyModel.ts [hunk 5]
```diff
@@ -319,21 +328,25 @@ export class DocumentBodyModel extends DocumentBodyModelSimple {
...
             }
+
             return this.getParagraphByTree(children, insertIndex);
         }
+
         return null;
     }
 
-    foreachTop(
+    private foreachTop(
```
Comment (wzhudev) [inline packages/core/src/Docs/Domain/DocumentBodyModel.ts:349]: A good habit to narrow interfaces! 👍

**Review by wzhudev — CHANGES_REQUESTED** [review]
The selection rendering issue needs to be addressed.

**Comment by wzhudev** [comment]
The toggling bold shortcut could be implemented in this PR as well?

File: packages/core/src/Docs/Domain/DocumentBodyModel.ts (hunk 5 above)
Comment (wzhudev) [reply packages/core/src/Docs/Domain/DocumentBodyModel.ts:349]: The name is better to be `forEachTop` like the native `forEach`.

File: packages/base-docs/src/Controller/doc-render.controller.ts (hunk 1 above)
Comment (DR-Univer) [reply packages/base-docs/src/Controller/doc-render.controller.ts:180]: recalculateSizeBySkeleton，里的resize会触发渲染

File: packages/ui-plugin-docs/src/Controller/menu/menu.ts [hunk 6]
```diff
@@ -1,11 +1,11 @@
-import { BreakLineCommand, DeleteLeftCommand } from '@univerjs/base-docs';
+import { BreakLineCommand, SetInlineFormatBoldCommand } from '@univerjs/base-docs';
 import { IMenuButtonItem, MenuItemType, MenuPosition } from '@univerjs/base-ui';
 import { IAccessor } from '@wendellhu/redi';
 
-// TODO@Dushusir: use for test, change id later
+// TODO @Dushusir: use for test, change id later
 export function BoldMenuItemFactory(accessor: IAccessor): IMenuButtonItem {
     return {
-        id: DeleteLeftCommand.id,
```
Comment (wzhudev) [inline packages/ui-plugin-docs/src/Controller/menu/menu.ts:8]: It should have a `activated$` observable.

File: packages/base-docs/src/Controller/doc-render.controller.ts (hunk 1 above)
Comment (DR-Univer) [reply packages/base-docs/src/Controller/doc-render.controller.ts:180]: skeleton.calculate()要慎用，多次计算，后续会带来很大开销。normal-input.controller.ts 里输入时已经出发了calculate

Comment (wzhudev) [reply packages/base-docs/src/Controller/doc-render.controller.ts:180]: @DR-Univer 可以做一个性能监控，统计重算排版布局的次数

**Comment by Jocs** [comment]
> The toggling bold shortcut could be implemented in this PR as well?

@wzhudev I tend to put it in a different PR, This PR is mainly to run through the process of inline formats.

File: packages/base-docs/src/Controller/doc-render.controller.ts (hunk 1 above)
Comment (Jocs) [reply packages/base-docs/src/Controller/doc-render.controller.ts:180]: 选取的重新渲染还没做，可以加上去

Comment (Jocs) [reply packages/base-docs/src/Controller/doc-render.controller.ts:180]: > skeleton.calculate()要慎用，多次计算，后续会带来很大开销。normal-input.controller.ts 里输入时已经出发了calculate

这儿要不要做一个区分，只是某些 mutation 出发 calculate() ？

File: packages/base-docs/src/Basics/commonParameter.ts (hunk 2 above)
Comment (Jocs) [reply packages/base-docs/src/Basics/commonParameter.ts:1]: 额，是的，MemoryCursor A HA?

File: packages/base-docs/src/commands/mutations/core-editing.mutation.ts (hunk 4 above)
Comment (Jocs) [reply packages/base-docs/src/commands/mutations/core-editing.mutation.ts:74]: 没太明白什么意思？要不解释下？

File: packages/core/src/Docs/Domain/DocumentBodyModel.ts (hunk 5 above)
Comment (Jocs) [reply packages/core/src/Docs/Domain/DocumentBodyModel.ts:349]: +1

File: packages/ui-plugin-docs/src/Controller/menu/menu.ts (hunk 6 above)
Comment (Jocs) [reply packages/ui-plugin-docs/src/Controller/menu/menu.ts:8]: 这个没有实现，选取改变应该触发菜单状态变化，这个应该放在另外一个 PR 中？

File: packages/base-docs/src/Controller/inline-format.controller.ts (hunk 3 above)
Comment (Jocs) [reply packages/base-docs/src/Controller/inline-format.controller.ts:22]: +1

File: packages/ui-plugin-docs/src/Controller/menu/menu.ts (hunk 6 above)
Comment (wzhudev) [reply packages/ui-plugin-docs/src/Controller/menu/menu.ts:8]: 也 OK

**Comment by Jocs** [comment]
> And there is another issue: the toggle bold button in the toolbar should be activated if the selected texts are in bold. @jikkai The sheet demo seems abnormal on this as well. Could you please take a look at it?

@wzhudev This should not have been implemented yet, I wrote a method to determine the format in selections, which can be used later.

File: packages/base-docs/src/commands/mutations/core-editing.mutation.ts (hunk 4 above)
Comment (wzhudev) [reply packages/base-docs/src/commands/mutations/core-editing.mutation.ts:74]: Mutation 一但生成，除了做 transform 应该不可变，这里在应用的时候还会更改原来的 mutation 有点怪

**Comment by wzhudev** [comment]
> > And there is another issue: the toggle bold button in the toolbar should be activated if the selected texts are in bold. @jikkai The sheet demo seems abnormal on this as well. Could you please take a look at it?
> 
> @wzhudev This has not been implemented yet, I wrote a method to determine the format in selections, which can be used later.

OK. However, we should fix that for Univer sheet. 💯

File: packages/base-docs/src/commands/mutations/core-editing.mutation.ts (hunk 4 above)
Comment (Jocs) [reply packages/base-docs/src/commands/mutations/core-editing.mutation.ts:74]: 是的，所以这个深拷贝了一下，后面（不是这个 PR）会把下面几个函数修改成纯函数，就可以把这个深拷贝去掉了。

**Comment by jikkai** [comment]
> The following issues need to be addressed before we merge this.
> 
> And there is another issue: the toggle bold button in the toolbar should be activated if the selected texts are in bold. @jikkai The sheet demo seems abnormal on this as well. Could you please take a look at it?

fixed in #364

File: packages/base-docs/src/Controller/doc-render.controller.ts (hunk 1 above)
Comment (Jocs) [reply packages/base-docs/src/Controller/doc-render.controller.ts:180]: > 选取的重新渲染还没做，可以加上去

这个加上了

**Review by wzhudev — APPROVED** [review]

**Pull request merged**